swarm/src/behaviour.rs

```diff
@@ -288,188 +288,12 @@ pub enum NetworkBehaviourAction<TOutEvent, TInEvent> {
 
     /// Instructs the swarm to start a dial.
     ///
-    /// On success, [`NetworkBehaviour::inject_connection_established`] is invoked.
-    /// On failure, [`NetworkBehaviour::inject_dial_failure`] is invoked.
+    /// On success, [`NetworkBehaviour::new_handler`] is called with the [`PeerId`] and [`ConnectedPoint`]
+    /// of the newly established connection.
     ///
-    /// Note that the provided handler is returned to the [`NetworkBehaviour`] on connection failure
-    /// and connection closing. Thus it can be used to carry state, which otherwise would have to be
-    /// tracked in the [`NetworkBehaviour`] itself. E.g. a message destined to an unconnected peer
-    /// can be included in the handler, and thus directly send on connection success or extracted by
-    /// the [`NetworkBehaviour`] on connection failure.
+    /// Once the handler is constructed, [`NetworkBehaviour::inject_connection_established`] will be called.
     ///
-    /// # Example carrying state in the handler
```

Comment:
Should we replace this example? Now that `new_handler` gives you a `PeerId`, I think it is quite easy to realize that you can store data for a future connection in the behaviour and just pass it into the handler here.

Reply:
> * `NetworkBehaviour`s can track state for future connections within themselves (indexed by `PeerId`) and pass it into the `ConnectionHandler` upon construction. This removes the need for passing along a handler in `NetworkBehaviourAction::Dial`.

Say there are two `NetworkBehaviour` implementations, each requesting a new connection to a new peer. In such case, when `NetworkBehaviour::inject_connection_established` is called, neither of them knows whether this new connection corresponds to their dial request.

We could still allow the user to provide user data via `DialOpts`. For the case of `Swarm::dial` and `NetworkBehaviour::inject_dial_failure` we could wrap this in an `Option`.

I am not sure whether we need to design for the above race condition. Your proposal might be just fine. What do folks think?

Reply:
In `ipfs-embed` I validate advertised peer addresses by dialling them — and closing such connections upon success. This is only reliably possible if I can detect whether the connection resulted from my own dialling attempt. OTOH, another cause may have resulted in dialling that same peer with that same address for other reasons, which would then possibly have said “nah, attempt is already underway”. But all of this would happen _in addition to existing connections_, so it might not be a huge problem.

Reply:
Do you need to actively close them? Assuming reasonable `keep_alive` implementations of your `ConnectionHandler`s, the connection should get closed automatically if it is not in use.

> OTOH, another cause may have resulted in dialling that same peer with that same address for other reasons, which would then possibly have said “nah, attempt is already underway”.

Can this be expressed with [`PeerCondition::NotDialing`](https://docs.rs/libp2p/0.49.0/libp2p/swarm/dial_opts/enum.PeerCondition.html#variant.NotDialing)?

Note that `new_handler` also gives you access to `ConnectedPoint`. So in addition to storing data in your behaviour based on `PeerId`, you could also index it by the dialed address. `new_handler` being called for an address you wanted to validate IS the validation that a connection was made to this address. If you have a dedicated `AddressValidationBehaviour`, that behaviour could then straight up denying that connection which instantly closes it again.